File: sdks/python-mcp/python/solvapay_mcp/asgi/mcp_resolve_auth.py
"""Host-side `mcpResolveAuth`: one JSON contract for MCP bearer decisions."""

from __future__ import annotations

from collections.abc import Mapping
from typing import TypedDict, TypeGuard

from solvapay_mcp.server.helpers import _invoke


class McpResolveAuthParams(TypedDict, total=False):
    rpcMethod: str | None
    authHeader: str | None
    authMode: str
    publicBaseUrl: str
    mcpPath: str
    jsonRpcId: str | int | None
    hs256Secret: str
    jwksJson: object


def _is_record(value: object) -> TypeGuard[dict[str, object]]:
    return isinstance(value, dict)


async def mcp_resolve_auth(params: Mapping[str, object], client: object) -> dict[str, object]:
    if client is None:
        raise RuntimeError("oauth_client is required for mcpResolveAuth")
    envelope = await _invoke(client, "mcp_resolve_auth", dict(params))
    if not _is_record(envelope):
        raise RuntimeError("mcpResolveAuth returned a non-object envelope")
    kind = envelope.get("kind")
    if kind not in {"allow", "challenge", "error"}:
        raise RuntimeError(f"mcpResolveAuth returned unexpected kind: {kind!r}")
    return envelope

File: sdks/typescript/mcp/src/internal/mcp-resolve-auth.ts
/**
 * Host-side `mcpResolveAuth`: one JSON contract for MCP bearer decisions.
 */

export type McpResolveAuthParams = {
  rpcMethod?: string
  authHeader?: string | null
  authMode?: string
  publicBaseUrl: string
  mcpPath?: string
  jsonRpcId?: string | number | null
  hs256Secret?: string
  jwksJson?: unknown
}

export type McpResolveAuthClient = {
  mcpResolveAuth: (params: McpResolveAuthParams) => Promise<unknown>
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return value !== null && typeof value === 'object'
}

export async function mcpResolveAuth(
  params: McpResolveAuthParams,
  client: McpResolveAuthClient,
): Promise<Record<string, unknown>> {
  const envelope = await client.mcpResolveAuth(params)
  if (!isRecord(envelope)) {
    throw new Error('mcpResolveAuth returned a non-object envelope')
  }
  const kind = envelope.kind
  if (kind !== 'allow' && kind !== 'challenge' && kind !== 'error') {
    throw new Error(`mcpResolveAuth returned unexpected kind: ${String(kind)}`)
  }
  return envelope
}

export function requireResolveAuthClient(
  client: McpResolveAuthClient | null | undefined,
): McpResolveAuthClient {
  if (client == null) {
    throw new Error(
      'oauthClient is required; pass the SolvaPay API client that implements mcpResolveAuth',
    )
  }
  return client
}
